netebpfext/net_ebpf_ext_xdp.c

```diff
@@ -45,28 +40,219 @@ DEFINE_GUID(EBPF_ATTACH_TYPE_XDP, 0x85e0d8ef, 0x579e, 0x4931, 0xb0, 0x72, 0x8e,
+    net_xdp_ctx->data = packet_buffer;
+    net_xdp_ctx->data_end = packet_buffer + cloned_net_buffer_length;
+
+    // Create a MDL with the packet buffer.
+    mdl_chain = IoAllocateMdl(packet_buffer, cloned_net_buffer_length, FALSE, FALSE, NULL);
+    if (mdl_chain == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+    MmBuildMdlForNonPagedPool(mdl_chain);
+
+    // Now allocate the cloned NBL using this MDL chain.
+    status = FwpsAllocateNetBufferAndNetBufferList(
+        _net_ebpf_ext_nbl_pool_handle, 0, 0, mdl_chain, 0, cloned_net_buffer_length, &new_nbl);
+    if (!NT_SUCCESS(status))
+        goto Exit;
+    mdl_chain = NULL;
+    packet_buffer = NULL;
+
+    // Set the new NBL as the cloned NBL in XDP context, after disposing any previous clones.
+    if (net_xdp_ctx->cloned_nbl != NULL)
+        _net_ebpf_ext_free_nbl(net_xdp_ctx->cloned_nbl);
+    net_xdp_ctx->cloned_nbl = new_nbl;
+
+Exit:
+    if (mdl_chain != NULL)
+        IoFreeMdl(mdl_chain);
+    if (packet_buffer != NULL)
+        ExFreePool(packet_buffer);
+    return status;
+}
+
+static int
+_net_ebpf_xdp_adjust_head(_Inout_ xdp_md_t* ctx, int delta)
+{
+    int return_value = 0;
+    NDIS_STATUS ndis_status = NDIS_STATUS_SUCCESS;
+    net_ebpf_xdp_md_t* net_xdp_ctx = (net_ebpf_xdp_md_t*)ctx;
+    NET_BUFFER_LIST* nbl = NULL;
+    NET_BUFFER* net_buffer = NULL;
+    uint8_t* packet_buffer = NULL;
+
+    // Either original or cloned NBL must be present.
+    if ((net_xdp_ctx->original_nbl == NULL) && (net_xdp_ctx->cloned_nbl == NULL)) {
+        return_value = -1;
+        goto Exit;
+    }
+
+    nbl = (net_xdp_ctx->cloned_nbl != NULL) ? net_xdp_ctx->cloned_nbl : net_xdp_ctx->original_nbl;
+    ASSERT(nbl != NULL);
+    net_buffer = NET_BUFFER_LIST_FIRST_NB(nbl);
+
+    if (delta == 0)
+        // Nothing to do.
+        goto Exit;
+    if (delta < 0) {
+        uint32_t absolute_delta = -delta;
+        ndis_status = NdisRetreatNetBufferDataStart(net_buffer, absolute_delta, 0, NULL);
+        if (ndis_status != NDIS_STATUS_SUCCESS) {
+            return_value = -1;
+            goto Exit;
+        }
+        packet_buffer = NdisGetDataBuffer(net_buffer, net_buffer->DataLength, NULL, 1, 0);
+        if (packet_buffer != NULL)
+            net_xdp_ctx->data = packet_buffer;
+        else {
+            // Data in net_buffer not contiguous.
+            // Restore net_buffer.
+            NdisAdvanceNetBufferDataStart(net_buffer, absolute_delta, TRUE, NULL);
+            // Allocate a cloned NBL with contiguous data.
+            _net_ebpf_ext_allocate_cloned_nbl(net_xdp_ctx, absolute_delta);
+        }
+    } else {
+        // delta > 0.
+        NdisAdvanceNetBufferDataStart(net_buffer, delta, FALSE, NULL);
+        packet_buffer = NdisGetDataBuffer(net_buffer, net_buffer->DataLength, NULL, 1, 0);
+        ASSERT(packet_buffer != NULL);
+        net_xdp_ctx->data = packet_buffer;
+    }
+
+Exit:
+    return return_value;
+}
+
+void
```

Comment:
```suggestion
static void
```
functions starting with `_` should be static